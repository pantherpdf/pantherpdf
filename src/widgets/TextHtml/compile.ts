import { listOfAdjusts } from '../formulaAdjust';
import { CompileHelper, WidgetItem } from '../types';
import type { TextHtmlCompiled, TextHtmlData } from './TextHtml';

export default async function compile(
  item: WidgetItem,
  helper: CompileHelper,
): Promise<TextHtmlCompiled> {
  const dt = item as TextHtmlData;
  // combine parts
  let html = '';
  for (const part of dt.value) {
    const type = part.type;
    switch (type) {
      case 'html': {
        html += part.value;
        break;
      }
      case 'formula': {
        let value = await helper.evalFormula(part);
        if (part.adjust && part.adjust.length > 0) {
          const adjustObj = listOfAdjusts.find(x => x.id === part.adjust);
          if (!adjustObj) {
            throw new Error(`Unknown adjust ${part.adjust}`);
          }
          value = adjustObj.func(value, []);
        }
        html += String(value);
        break;
      }
      default: {
        const exhaustiveCheck: never = type;
        throw new Error(`Unknown data type: ${exhaustiveCheck}`);
      }
    }
  }
  return {
    type: dt.type,
    value: html,
    font: dt.font,
  };
}
