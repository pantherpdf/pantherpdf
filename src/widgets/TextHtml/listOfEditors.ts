/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import { idCmp } from '../../editor/childrenMgmt';
import type Editor from './Editor';

const listOfEditors: Editor[] = [];
const listOfSelectionCallbacks: (() => void)[] = [];

function editorGet(wid: number[]): Editor | undefined {
  return listOfEditors.find(edt => idCmp(edt.props.wid, wid));
}

if (typeof window !== 'undefined' && window.document) {
  window.document.addEventListener('selectionchange', () => {
    const s = window.getSelection();
    const anchorNode = s ? s.anchorNode : null;
    const anchorOffset = s ? s.anchorOffset : 0;
    function getEditorByEditor(node: Node) {
      // keep function outside of loop
      // https://eslint.org/docs/rules/no-loop-func
      return listOfEditors.find(edt => edt.elementRef === node);
    }
    let editor: Editor | undefined = undefined;
    let node = anchorNode;
    while (node) {
      // check if editor
      editor = getEditorByEditor(node);
      if (editor) {
        break;
      }
      // next parent
      node = node.parentNode;
    }
    // reset all
    for (const edt of listOfEditors) {
      if (edt === editor) {
        edt.saveCaret(anchorNode, anchorOffset);
      }
    }
    // callbacks
    for (const cb of listOfSelectionCallbacks) {
      cb();
    }
  });
}

export { listOfEditors, listOfSelectionCallbacks, editorGet };
