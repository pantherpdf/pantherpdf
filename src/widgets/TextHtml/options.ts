/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import type { TransName } from '../../translation';

const tagType = 'data';
const name: TransName = { en: 'Text', sl: 'Besedilo' };

export { tagType, name };
