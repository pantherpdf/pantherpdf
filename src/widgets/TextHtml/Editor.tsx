/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import React, { Component, CSSProperties } from 'react';
import { styled } from '@mui/material/styles';
import { listOfEditors, listOfSelectionCallbacks } from './listOfEditors';
import {
  getNodeFromPid,
  getPidFromNode,
  getTagFromSelection,
  insertTag,
} from './helpers';

interface EditorProps {
  style?: CSSProperties;
  wid: number[];
  value: string;
  onChange: (val: string) => void;
  active: boolean;
}
interface EditorState {}

// use class component instead of function - cleaner design
//  * no need to use useRef()
//  * access to shouldComponentUpdate()
export default class Editor extends Component<EditorProps, EditorState> {
  // reference to <div>
  elementRef: HTMLDivElement | null;
  // to know when change is comming from here and when from outside
  currentValueFromProps: string;
  // timeout for callback to save changes
  timerUpdate: number;
  // selection - caret position
  selectedNode: number[];
  selectedOffset: number;

  constructor(props: EditorProps) {
    super(props);
    this.state = {};

    this.elementRef = null;
    this.currentValueFromProps = props.value;
    this.timerUpdate = 0;

    this.selectedNode = [];
    this.selectedOffset = 0;
  }

  componentDidMount() {
    listOfEditors.push(this);
  }

  componentWillUnmount() {
    if (this.timerUpdate) {
      this.sendChanges(true);
    }
    const idx = listOfEditors.indexOf(this);
    if (idx !== -1) {
      listOfEditors.splice(idx, 1);
    }
  }

  applyUpdates(el: HTMLDivElement, props: EditorProps, force: boolean) {
    // data-wid
    const widStr = props.wid.map(x => String(x)).join(',');
    el.setAttribute('data-wid', widStr);

    // style
    el.removeAttribute('style');
    if (props.style) {
      Object.assign(el.style, props.style);
    }
    // padding to separate content from outline (when it has focus), to show cursor
    el.style.padding = '1px 3px';

    // html
    if (props.value !== this.currentValueFromProps || force) {
      el.innerHTML = props.value;
      this.currentValueFromProps = props.value;
      this.saveCaret(null, 0);
    }
  }

  setElementRef(el: HTMLDivElement | null) {
    if (el === this.elementRef) {
      return;
    }
    if (el) {
      // update <div>
      this.applyUpdates(el, this.props, true);
    }
    this.elementRef = el;
  }

  shouldComponentUpdate(nextProps: EditorProps): boolean {
    if (this.elementRef) {
      this.applyUpdates(this.elementRef, nextProps, false);
    }
    // never re-render
    return false;
  }

  saveCaret(node: Node | null, offset: number): void {
    const newNode =
      node && this.elementRef ? getPidFromNode(node, this.elementRef) : [];
    // is same?
    if (this.selectedNode === newNode && this.selectedOffset === offset) {
      return;
    }
    // update
    this.selectedNode = newNode;
    this.selectedOffset = offset;
    for (const cb of listOfSelectionCallbacks) {
      cb();
    }
  }

  editorInput(e: React.FormEvent<HTMLDivElement>) {
    e.stopPropagation();
    // set timer to update after few seconds
    if (this.timerUpdate > 0) {
      clearTimeout(this.timerUpdate);
      this.timerUpdate = 0;
    }
    const tm = window.setTimeout(this.sendChanges.bind(this), 3000, false);
    this.timerUpdate = tm;
  }

  sendChanges(cleanHtml: boolean) {
    // callback from timer
    if (this.timerUpdate) {
      clearTimeout(this.timerUpdate);
      this.timerUpdate = 0;
    }
    if (!this.elementRef) {
      return;
    }
    // change
    const val = this.elementRef.innerHTML;
    if (val === this.currentValueFromProps) {
      return;
    }
    if (cleanHtml) {
      // todo sanitize
      // todo remove empty tag
      // todo remove nbsp at end of line
      // todo remove tags like <span style="white-space: nowrap;">new real text</span>
      // todo remove style padding, margin, ...
      if (val === this.currentValueFromProps) {
        return;
      }
    }
    this.currentValueFromProps = val;
    this.props.onChange(val);
  }

  editorKeyDown(e: React.KeyboardEvent<HTMLDivElement>): void {
    e.stopPropagation();
    // when user presses enter, manually add new line to prevent adding new line inside tag
    if (e.key !== 'Enter') {
      return;
    }
    if (!this.elementRef) {
      return;
    }
    const node = getNodeFromPid(this.selectedNode, this.elementRef);
    if (!node) {
      return;
    }
    const btn = getTagFromSelection(node, this.elementRef);
    if (!btn || !btn.parentElement) {
      return;
    }
    // add new line after <?>
    e.preventDefault();
    const el = document.createElement('div');
    const nodeTxt = document.createTextNode('\u00a0');
    el.appendChild(nodeTxt);
    btn.parentElement.insertBefore(el, btn.nextSibling);
    // move caret
    const range = document.createRange();
    range.setStart(nodeTxt, 0);
    range.setEnd(nodeTxt, 1);
    range.collapse(true);
    const s = window.getSelection();
    if (s) {
      s.removeAllRanges();
      s.addRange(range);
    }
  }

  // hack for safari
  // without removing draggable, text selection doesnt work
  // https://stackoverflow.com/questions/6399131/html5-draggable-and-contenteditable-not-working-together
  parentsRemoveDraggable(el: HTMLElement | null) {
    while (el) {
      if (el.draggable) {
        el.draggable = false;
        el.setAttribute('data-draggable', '1');
      }
      el = el.parentElement;
    }
  }

  parentsRestoreDraggable(el: HTMLElement | null) {
    while (el) {
      if (el.getAttribute('data-draggable') === '1') {
        el.draggable = true;
        el.removeAttribute('data-draggable');
      }
      el = el.parentElement;
    }
  }

  dropTag(e: React.DragEvent<HTMLDivElement>) {
    // https://developer.mozilla.org/en-US/docs/Web/API/Document/caretRangeFromPoint
    // https://jsfiddle.net/fpkjbech/
    e.preventDefault();
    /* eslint-disable @typescript-eslint/no-explicit-any */
    const document_ = document as any;
    const e_ = e as any;
    /* eslint-enable @typescript-eslint/no-explicit-any */
    let node: Node;
    let offset: number;
    e.currentTarget.focus();
    if (document_.caretRangeFromPoint) {
      const range = document.caretRangeFromPoint(e.clientX, e.clientY);
      if (!range) {
        return;
      }
      node = range.startContainer;
      offset = range.startOffset;
    }
    //
    else if (document_.caretPositionFromPoint) {
      const range = document_.caretPositionFromPoint(e.clientX, e.clientY);
      if (!range) {
        return;
      }
      node = range.offsetNode;
      offset = range.offset;
    }
    //
    else if (e_.rangeParent) {
      const range = document.createRange();
      range.setStart(e_.rangeParent, e_.rangeOffset);
      node = range.startContainer;
      offset = range.startOffset;
    }
    //
    else {
      // not supported
      return;
    }
    // update selection
    if (!node || !this.elementRef) {
      return;
    }
    this.selectedNode = getPidFromNode(node, this.elementRef);
    this.selectedOffset = offset;
    for (const cb of listOfSelectionCallbacks) {
      cb();
    }
    // editor
    const widStr = e.currentTarget.getAttribute('data-wid');
    if (!widStr || widStr.length === 0) {
      return;
    }
    const wid = widStr.split(',').map(x => parseInt(x));
    // insert
    insertTag(wid);
  }

  render() {
    return (
      <TextHtmlEditor
        ref={this.setElementRef.bind(this)}
        contentEditable
        draggable
        onDragStart={e => {
          e.preventDefault();
          e.stopPropagation();
        }}
        onKeyDown={this.editorKeyDown.bind(this)}
        onInput={this.editorInput.bind(this)}
        onFocus={e => {
          this.parentsRemoveDraggable(e.currentTarget);
        }}
        onBlur={e => {
          this.parentsRestoreDraggable(e.currentTarget);
          this.sendChanges(true);
        }}
        onDrop={e => {
          const dt = e.dataTransfer.getData('text/plain');
          if (dt === 'insert-tag') {
            this.dropTag(e);
          }
        }}
      />
    );
  }
}

const TextHtmlEditor = styled('div')({
  borderRadius: '0.15rem',

  '&:focus-visible': {
    boxShadow: '0 0 3pt 2pt rgba(100,149,237,0.7)',
    outline: 'none',
  },

  '& data': {
    display: 'inline-block',
    cursor: 'pointer',
    outline: 'none',
    border: '1px solid #949ea8',
    borderRadius: '0.2rem',
    color: 'inherit',
    padding: '0 0.05rem',
    backgroundColor: 'inherit',
    fontWeight: 'inherit',
    fontSize: 'inherit',
    fontFamily: 'inherit',
    fontStyle: 'inherit',
    textAlign: 'inherit',
    maxWidth: '100%',
    overflow: 'hidden',
    whiteSpace: 'nowrap',
    verticalAlign: 'top',
  },
});
