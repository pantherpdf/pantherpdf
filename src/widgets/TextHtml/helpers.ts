/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import { editorGet } from './listOfEditors';
import { tagType } from './options';

// return first parent tag
// node is usualy text
export function getTagFromSelection(
  node: Node,
  editor: HTMLElement,
): HTMLElement | null {
  if (node === editor) {
    return null;
  }
  if (node.nodeName.toLowerCase() === tagType.toLowerCase()) {
    return node as HTMLElement;
  }
  if (node.parentNode) {
    return getTagFromSelection(node.parentNode, editor);
  }
  return null;
}

// make number path from editor to child node
export function getPidFromNode(node: Node, editor: HTMLElement): number[] {
  const arr: number[] = [];
  while (node !== editor) {
    const parent = node.parentNode;
    if (!parent) {
      throw new Error('Cant find pid 1');
    }
    let found = false;
    for (let i = 0; i < parent.childNodes.length; ++i) {
      if (parent.childNodes[i] === node) {
        arr.push(i);
        found = true;
      }
    }
    if (!found) {
      throw new Error('Cant find pid 2');
    }
    node = parent;
  }
  arr.reverse();
  return arr;
}

// return nth child of editor
export function getNodeFromPid(
  ids: number[],
  editor: HTMLElement,
): Node | null {
  let base: Node = editor;
  for (const idx of ids) {
    if (idx >= base.childNodes.length) {
      return null;
    }
    base = base.childNodes[idx];
  }
  return base;
}

export function insertTag(wid: number[]) {
  const editor = editorGet(wid);
  if (!editor || !editor.elementRef) {
    return;
  }
  const selectedNode = getNodeFromPid(editor.selectedNode, editor.elementRef);
  if (!selectedNode || !selectedNode.parentNode) {
    return;
  }
  const parentNode = selectedNode.parentNode;

  // is inside tag?
  {
    let el: Node | null = selectedNode;
    while (el) {
      if (el.nodeName.toLowerCase() === tagType.toLowerCase()) {
        return;
      }
      el = el.parentNode;
    }
  }

  const btn = document.createElement(tagType.toUpperCase());
  btn.setAttribute('data-adjust', '');
  const btnValue = 'data';
  const btnTextNode = document.createTextNode(btnValue);
  btn.appendChild(btnTextNode);

  if (selectedNode.nodeName === '#text') {
    const txtBefore =
      selectedNode.nodeValue?.substring(0, editor.selectedOffset) || '';
    const txtAfter =
      (selectedNode.nodeValue?.substring(editor.selectedOffset) || '') +
      '\u00a0'; // &nbsp;
    const nextSib = selectedNode.nextSibling;
    selectedNode.nodeValue = txtBefore;
    parentNode.insertBefore(btn, nextSib);
    parentNode.insertBefore(document.createTextNode(txtAfter), nextSib);
  }
  //
  else if (selectedNode === editor.elementRef) {
    selectedNode.appendChild(btn);
    selectedNode.appendChild(document.createTextNode('\u00a0'));
  }
  //
  else {
    const nextSib = selectedNode.nextSibling;
    parentNode.insertBefore(btn, nextSib);
    parentNode.insertBefore(document.createTextNode('\u00a0'), nextSib);
  }

  // select
  const range = document.createRange();
  range.setStart(btnTextNode, 0);
  range.setEnd(btnTextNode, btnValue.length);
  range.collapse(true);
  const s = window.getSelection();
  if (s) {
    s.removeAllRanges();
    s.addRange(range);
  }
}

export function escapeHtml(unsafe: string): string {
  return unsafe
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;');
}
