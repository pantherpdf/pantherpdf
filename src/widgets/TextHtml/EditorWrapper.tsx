import React from 'react';
import { LoadGoogleFontCss } from '../GoogleFonts';
import { PropertyFontExtractStyle, PropertyFontGenCss } from '../PropertyFont';
import WidgetEditorName from '../WidgetEditorName';
import { WidgetEditorProps } from '../types';
import Editor from './Editor';
import { type TextHtmlData } from './TextHtml';
import {
  valueInternalFromEditor,
  valueInternalToEditor,
} from './internalRepresentation';
import { idCmp } from '../../editor/childrenMgmt';
import { name } from './options';

export default function EditorWrapper(props: WidgetEditorProps) {
  const item = props.item as TextHtmlData;
  const css = PropertyFontGenCss(item.font);
  css.minHeight = '20px';
  const font2 = PropertyFontExtractStyle(item.font);
  if (font2) {
    LoadGoogleFontCss(font2);
  }

  return (
    <WidgetEditorName
      {...props}
      name={name}
      visible={!props.selected || !idCmp(props.selected, props.wid)}
    >
      <div
        onClick={e => {
          e.stopPropagation();
          const isSelected =
            JSON.stringify(props.selected) === JSON.stringify(props.wid);
          if (!isSelected) {
            props.setSelected(props.wid);
          }
        }}
      >
        <Editor
          wid={props.wid}
          value={valueInternalToEditor(item.value)}
          onChange={val =>
            props.setItem({ ...item, value: valueInternalFromEditor(val) })
          }
          style={css}
          active={!!props.selected && idCmp(props.wid, props.selected)}
        />
      </div>
    </WidgetEditorName>
  );
}
