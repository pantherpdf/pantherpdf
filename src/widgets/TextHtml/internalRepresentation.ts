/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

import { FormulaObject } from '../../types';
import { Element_, parse, extractText } from '../HtmlParser';
import { escapeHtml } from './helpers';
import { tagType } from './options';

export type ValueHtml = { type: 'html'; value: string };
export type ValueFormula = FormulaObject & { type: 'formula'; adjust?: string };
export type TextHtmlDataValue = ValueHtml | ValueFormula;

export function valueInternalToEditor(value: TextHtmlDataValue[]): string {
  let html = '';
  for (const part of value) {
    const type = part.type;
    switch (type) {
      case 'html': {
        html += part.value;
        break;
      }
      case 'formula': {
        html += `<${tagType}`;
        if (part.adjust) {
          html += ` data-adjust="${escapeHtml(part.adjust)}"`;
        }
        html += `>${escapeHtml(part.formula)}</${tagType}>`;
        break;
      }
      default: {
        const exhaustiveCheck: never = type;
        throw new Error(`Unknown data type: ${exhaustiveCheck}`);
      }
    }
  }
  return html;
}

export function extractTag(html: string) {
  function process(el: Element_): Element_[] {
    for (const el2 of el.children) {
      if (el2.name === '#text' && el2.textContent.trim().length > 0) {
        return [el];
      }
      const sub = process(el2);
      if (sub.length > 0) {
        return [el, ...sub];
      }
    }
    return [];
  }

  const r = parse(html);
  if (r.length !== 1) {
    throw new Error('Bad element');
  }
  const adjust = r[0].attributes['data-adjust'] || '';

  const tags2 = process(r[0]);
  if (tags2.length > 0) {
    // remove el
    tags2.splice(0, 1);
  }
  const tags = tags2.map(el => ({ name: el.name, attributes: el.attributes }));
  return {
    formula: extractText(r[0]),
    adjust: adjust || undefined,
    tags,
  };
}

export function valueInternalFromEditor(html: string): TextHtmlDataValue[] {
  if (html.indexOf('<!--') !== -1) {
    throw new Error('Comments not supported');
  }
  const arr: TextHtmlDataValue[] = [];
  const tagStart = '<' + tagType;
  const tagEnd = '</' + tagType + '>';
  let i = 0;
  while (i < html.length) {
    const i2 = html.indexOf(tagStart, i);
    if (i2 === -1) {
      break;
    }
    const i3 = html.indexOf(tagEnd, i2);
    if (i3 === -1) {
      break;
    }
    const html2 = html.substring(i, i2);
    if (html2.length > 0) {
      const lastVal = getOrAddHtmlValue(arr);
      lastVal.value += html2;
    }
    const { formula, adjust, tags } = extractTag(
      html.substring(i2, i3 + tagEnd.length),
    );
    if (
      tags.length > 0 &&
      (arr.length === 0 || arr[arr.length - 1].type !== 'html')
    ) {
      arr.push({ type: 'html', value: '' });
    }
    const lastVal = getOrAddHtmlValue(arr);
    for (const tag of tags) {
      lastVal.value += `<${tag.name}`;
      for (const [key, value] of Object.entries(tag.attributes)) {
        lastVal.value += ` ${key}`;
        if (value !== undefined) {
          lastVal.value += `="${escapeHtml(value)}"`;
        }
      }
      lastVal.value += `>`;
    }
    if (formula.length > 0) {
      arr.push({ type: 'formula', formula, adjust });
    }
    if (tags.length > 0) {
      const val: TextHtmlDataValue = { type: 'html', value: '' };
      for (const tag of [...tags].reverse()) {
        val.value += `</${tag.name}>`;
      }
      arr.push(val);
    }
    i = i3 + tagEnd.length;
  }
  if (i < html.length) {
    const lastVal = getOrAddHtmlValue(arr);
    lastVal.value += html.substring(i);
  }
  return arr;
}

function getOrAddHtmlValue(arr: TextHtmlDataValue[]): ValueHtml {
  if (arr.length > 0 && arr[arr.length - 1].type === 'html') {
    return arr[arr.length - 1] as ValueHtml;
  }
  const lastVal: ValueHtml = { type: 'html', value: '' };
  arr.push(lastVal);
  return lastVal;
}
