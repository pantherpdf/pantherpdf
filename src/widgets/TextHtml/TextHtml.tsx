/**
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

// manual tests:
// - widget not selected, try to drag it. It should not be draggable, however it should select text
// - type some text, then wait few sec to periodicaly save, the cursor should stay on same place
// - insert tag, delete nbsp space after tag and press enter

import React from 'react';
import type { WidgetItem, WidgetCompiled, Widget } from '../types';
import { PropertyFontGenCss, TFont } from '../PropertyFont';
import { faAlignLeft } from '@fortawesome/free-solid-svg-icons';
import trans from '../../translation';
import PropertiesEditor from './PropertiesEditor';
import { TextHtmlDataValue } from './internalRepresentation';
import { escapeHtml } from './helpers';
import { name } from './options';
import compile from './compile';
import EditorWrapper from './EditorWrapper';

export interface TextHtmlData extends WidgetItem {
  type: 'TextHtml';
  value: TextHtmlDataValue[];
  font: TFont;
}

export interface TextHtmlCompiled extends WidgetCompiled {
  type: 'TextHtml';
  value: string;
  font: TFont;
}

export const TextHtml: Widget = {
  id: 'TextHtml',
  name,
  icon: faAlignLeft,

  newItem: async (props): Promise<TextHtmlData> => {
    const valueTxt = `<div>${escapeHtml(
      trans('TextHtml initial value'),
    )}</div>`;
    const value: TextHtmlDataValue[] =
      props.report.widgets.length > 0
        ? [{ type: 'html', value: valueTxt }]
        : [];
    return {
      type: 'TextHtml',
      children: [],
      value,
      font: {},
    };
  },

  compile,

  Editor: EditorWrapper,

  Preview: function (props) {
    const item = props.item as TextHtmlCompiled;
    const css = PropertyFontGenCss(item.font);
    return <div style={css} dangerouslySetInnerHTML={{ __html: item.value }} />;
  },

  Properties: PropertiesEditor,

  canDrag: false,
};
