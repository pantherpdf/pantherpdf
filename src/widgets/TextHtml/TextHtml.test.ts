/**
 * @jest-environment node
 * @project PantherPDF Report Editor
 * @copyright Ignac Banic 2021-2024
 * @license MIT
 */

/* eslint-disable @typescript-eslint/no-explicit-any */

import { TextHtmlData, TextHtmlCompiled } from './TextHtml';
import {
  ReportForceWidgets,
  compileComponentTest,
  renderWidget,
} from '../../unitTestHelpers';
import renderToHtml from '../../data/renderToHtml';
import { sampleReport } from '../../editor/sampleReport';
import { defaultWidgets } from '../allWidgets';
import { extractTag, valueInternalFromEditor } from './internalRepresentation';
import { ApiEndpoints } from '../../types';
import { googleFontCssUrl } from '../GoogleFonts';
import compile from '../../data/compile';

test('parse TextHtml formula', async () => {
  const html = `aaa
<div>a<b style="color:white">b</b>c</div>
d<b id="&quot;">ef</b> <data data-adjust="123&gt;4"><i></i><b>d</b>ata.def</data><br/>
<div>ghi</div>
<p></p>`;

  const arr = valueInternalFromEditor(html);
  expect((arr[0] as any).value).toBe(`aaa
<div>a<b style="color:white">b</b>c</div>
d<b id="&quot;">ef</b> <b>`);
  expect((arr[1] as any).formula).toBe('data.def');
  expect((arr[1] as any).adjust).toBe('123>4');
  expect((arr[2] as any).value).toBe(`</b><br/>
<div>ghi</div>
<p></p>`);
});

test('extractTag', () => {
  const { formula, adjust, tags } = extractTag(
    '<data data-adjust="123&gt;8"><i></i><div><i></i><span data-abc="123" def>456</span></div></data>',
  );
  expect(formula).toBe('456');
  expect(adjust).toBe('123>8');
  expect(tags.length).toBe(2);
  expect(tags[0].name).toBe('div');
  expect(tags[1].name).toBe('span');
  expect(tags[1].attributes).toEqual({ 'data-abc': '123', def: undefined });
});

test('TextHtml 2', async () => {
  const dt: TextHtmlData = {
    type: 'TextHtml',
    value: valueInternalFromEditor('Hello <data><i></i><b>data.txt</b></data>'),
    children: [],
    font: {},
  };
  const data = { txt: '123' };
  const p2 = await compileComponentTest(dt, data);
  expect(p2).toBeTruthy();
  expect(p2.type).toBe('TextHtml');
  const p = p2 as TextHtmlCompiled;
  expect(p.value).toBe('Hello <b>123</b>');
});

test('TextHtml Filter', async () => {
  const dt: TextHtmlData = {
    type: 'TextHtml',
    value: valueInternalFromEditor(
      'Hello <data data-adjust="num, 2 dec">data.num</data>',
    ),
    children: [],
    font: {},
  };
  const data = { num: 123.123456789 };
  const p2 = await compileComponentTest(dt, data);
  expect(p2).toBeTruthy();
  expect(p2.type).toBe('TextHtml');
  const p = p2 as TextHtmlCompiled;
  expect(p.value).toBe('Hello 123.12');
});

test('TextHtml should render html', async () => {
  const dt: TextHtmlData = {
    type: 'TextHtml',
    value: valueInternalFromEditor('Hello <b>world</b>'),
    children: [],
    font: { size: '20px' },
  };
  const html = await renderWidget(dt);
  expect(html).toMatchSnapshot();
});

test('TextHtml should include google font', async () => {
  const report: ReportForceWidgets<TextHtmlData> = {
    ...sampleReport,
    widgets: [
      {
        type: 'TextHtml',
        value: valueInternalFromEditor('text'),
        children: [],
        font: { size: '20px', family: 'Lato' },
      },
    ],
  };
  const api: ApiEndpoints = {
    fonts: {
      list: [],
      getCssUrls: arr => {
        const url = googleFontCssUrl(arr);
        return url ? [url] : [];
      },
    },
  };
  const compiled = await compile(report, {}, defaultWidgets, api);
  const html = renderToHtml(compiled, defaultWidgets, api);
  expect(html).toContain(
    'https://fonts.googleapis.com/css2?family=Lato:ital,wght@0,400&display=swap',
  );
});
