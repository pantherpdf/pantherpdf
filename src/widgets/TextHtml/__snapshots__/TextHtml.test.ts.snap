// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextHtml should render html 1`] = `
<div style="font-size:20px">
  Hello
  <b>
    world
  </b>
</div>
`;
